### ISSUE TITLE
Divide main loop iteration into stages + define what these stages should do

### ISSUE THREAD

**Issue by nchashch** [amethyst/amethyst#238]
# Introduction
Recently a lot of work went into figuring out how specific parts of the engine
should work and writing code for them (@aceeri's work on configs, @torkleyy's work on
asset management, my work on parallellizing the renderer, @ebkalderon's renderer rewrite).

Now would be a good time to take a higher level look at the engine and to specify
where all the pieces should go and how they should fit together. So that everyone
is on the same page, and so that we have a more specific plan for future work.

Here is my take on amethyst's overall architecture, based on current design.
It isn't complete. It only addresses the main loop iteration. So it doesn't touch
configuration, startup, shutdown, and a lot of other things.
# Proposal
I think we should divide the main loop iteration into three stages: input,
process, and output. All the input would be read from the outside world at 
input stage and written to the `World`. At process stage `System`s
would be run, this stage can't talk to the outside world. At output stage
various "views" would read the `World` and put pixels on the screen/sound in the
headphones/etc.
# Why divide main loop iteration into stages?
## Cleaner API
By factoring out a separate input stage. It would be possible to take
raw input events, filter out all the unnecessary stuff, translate all the good
stuff into a convenient form (tailored to a specific game, because translation
code would live in `State` methods and would be written by engine users), and
put the translated input into the `World`.

Then at the process stage engine users won't have to care about handling raw
input. They would get a convenient representation (through custom or built in
`World` resources, like `InputHandler`, or `CharacterMovementStatus`). And they
would be able to pass through only the input data they need.

In order to put things on the screen/play sounds engine user would modify
special resources and components, which would then be read by "views" running
at output stage, and these views would put things on the screen/play sounds.
## Better granularity for debugging and profiling
Because input, processing, and output are isolated in separate stages it is
a bit easier to narrow down where the problem is. For example if there is some 
kind of graphics artifact, we would immediately know that the problem lives
somewhere in the graphics view. Or if there is some kind of problem with input
we would know that the problem happens at input stage.
# Stages of the main loop iteration
## Input stage (Can only read from the outside world, can both read and write the `World`)
At this stage `State` methods are run.
- Get things (entities, components, assets, events, network packets) and put them into the `World`.
- Load the `World` from scene description/save files (in `State::on_start` method).
- Do state transitions here (like quitting the game, going from main menu `State`
to gameplay `State`, pausing the game).
- Open question: do we serialize and save things here? Thinking of `State::on_stop` method.
### Get step
Engine user doesn't have to care about this step. All of these things are done
before `State` methods are run and the results are passed to the `State` methods.
- Catch all the events
- Retrieve and deserialize all the required entities and components
- Retrieve and load all the assets
- Receive and deserialize all incoming network traffic
### Put step
This step happens inside `State` methods, so the engine user controls how raw
input is translated into the state of `World` resources and components.
- Get all the relevant things acquired in the get step, and put them in the `World`.
## Process stage (Can neither read from nor write to the outside world, can both read and write the `World`)
Systems are dispatched here.

All the useful work specific to the game being developed should go here
(bookkeping is done at input stage, things that are the same for all games
are handled at output stage). And the code that does some particular bit of
useful work should be packaged as a `System`.

Game developers using the engine would communicate with the engine by
reading and writing `World` resources, components, and entities.

Input would go into specialized `World` resources (like `InputHandler`, `Time`,
`ScreenDimensions`, `NetworkReceiver`, `AssetProvider`). On input stage these
resources are updated. So they can be read at process stage.

`System`s run on process stage would update specialized output components
and resources (like `Renderable`, `SoundSource` components and `NetworkSender` resource).
On output stage various views would read these and turn them into pixels on the
screen/sound in headphones/packets sent by the wire.
### Graphics view: (something like `GfxDevice` in pre rewrite renderer)
1. Read all the graphics components and resources from the `World`.
2. Render the `World`.
3. Put pixels on the screen.
### Sound view:
1. Read all the sound components and resources from the `World`.
2. Do all the sound related work.
3. Put sound in player's speakers/headphones.
### Network "view":
1. Read all the networking related components and resources from the `World`.
2. Serialize all the things that need to be sent.
3. Send all the things to their respective destinations.

**Comment by torkleyy** [comment]
After reading this several times, I feel that the general idea of having stages is good. 
Here some thoughts:

* Graphics view, sound view and network view can be systems.

> For example if there is some
kind of graphics artifact, we would immediately know that the problem lives
somewhere in the graphics view.

* I don't really see how it makes debugging easier.

---

> It would be possible to take
raw input events, filter out all the unnecessary stuff, translate all the good
stuff into a convenient form (tailored to a specific game, because translation
code would live in State methods and would be written by engine users), and
put the translated input into the World.

* This can already be done in `State::on_events`, right?

---

I'm not really sure how you imagine the difference in the implementation exactly. More importantly, what would it mean for the user?

I don't think we should somehow hardcode and enforce this in the engine. I don't see how the division by the engine and the restrictions are necessary. But I may be misunderstanding your proposal, feel free to correct me @nchashch.

Instead, what came to my mind was that the ECS could offer stages so that you don't add single systems, but a whole group of systems. Consider the rendering system, which you want to run at the end; because you don't want to specify a dependency on every other system, it would make sense to put it into a stage and let the other things run before it. This would pretty much follow your definition of the "views" but it can be done the way the user wants it to be.

**Comment by nchashch** [comment]
> Graphics view, sound view and network view can be systems.

It would be nice, this way the views could be run concurrently and they could use the usual `specs` data parallelism mechanism (slide-rs/specs#120). Also amethyst's design this way would be "flatter", it would follow something like "everything that processes the `World` is a system" approach. Although it would require some way to ensure that all the view `System`s are run at the end.

> This can already be done in State::on_events, right?

Yes, in this issue I mostly described how the engine currently works:
1. `State` methods are run. - Input stage
2. `System`s are dispatched. - Process stage
3. `GfxDevice` reads and renders the `World`. - Output stage

I wanted to make it explicit, and give these stages names. Mostly to inform future design, figure out where things like audio view/network view/animation `System` would fit. Also I wanted to answer questions like "What should `State` methods do?"/"What should `System`s do?"/"What kind of API would the engine user see?".

> I'm not really sure how you imagine the difference in the implementation exactly. More importantly, what would it mean for the user?

I am not really proposing any changes to the existing code since it already follows what I have outlined in the issue. It is more about making all of these design choices explicit and providing some guidelines for future API design. I think we should create a document with a high-level description of the engine, outlining things like main loop stages/how input should work/how serialization and deserialization should work/how the output should work. It should help new contributors get up to speed more easily and for all the developers working on amethyst to have a shared coherent high-level understanding of the engine, I imagine it would become more and more important as the engine would grow and become less comprehensible by a single person.

Also, I think it would be nice to have an explicit high-level design of the engine hashed out and written down somewhere. So that before writing any code we could get an idea of where this code would go/when would it be run/what parts of the engine would it touch.

For example, consider the audio view. We can learn quite a lot about it before writing any code if we have a high-level picture of the engine:
1. It would be run at the end of the main loop iteration.
2. It would only read special audio `Component`s and resources from the `World` (it won't mutate the `World`).
3. No audio output code should be run at the input stage.
4. Engine users would talk to the audio view by reading and writing special audio `Component`s and resources, like `Speaker`/`Listener`/...

**Comment by torkleyy** [comment]
Okay, here is what I discussed with @ebkalderon and promised to document decades ago.

---

Amethyst aims to be a data-driven game engine, which means the behavior depends mostly on data, which allows for very nice and practical features (like hot reloading stuff).

Now, the same should be valid for our state machine. What I'm proposing is to reduce the `State` to initialization and cleanup of entities and components in the ECS world. So the ECS library (Specs) always just executes the same code, and a state change just means changing the data it operates on. This would also (like @nchashch's proposal) make it much clearer where code should go (in systems only) and we would utilize ECS more in order to allow for great parallelization and code reuse.

The stages you're asking for should be completely under control of the user by letting them build the `Dispatcher` (via `ApplicationBuilder`). As for separating these stages, see [this PR], which implemented barriers.

**Comment by nchashch** [comment]
TL;DR: I like the idea, it might even make sense to remove all code from `State`s and have them only point at serialized `World`s.

I like the idea of moving logic from `State`s to `System`s and only loading and unloading ECS data in `State` transitions. Then from the user's perspective answer to the question "What do `State`s do?" would be pretty simple: they load and unload `Entity`s, `Component`s, and `specs` resources.

And then a `State` would correspond to an initial state of the `World` - for example, a `Gameplay` `State` would correspond to "state of the `World` just at the moment after loading the save file". I think we will have to make `State`s serializable/de-serializable to make them more in line with data driven paradigm so that they would be mostly defined by the serialized data on disk and not by Rust code.

It might even make sense to define a standard serialization format that would be the same for all `State`s and a single serialization/de-serialization procedure that also would be shared by all `State`s. This way `State`s won't contain any code at all. Maybe a `State` would contain an asset identifier, pointing to a serialized `World`, then a state transition would consist of:
1. Unload the current `World`
2. Fetch the serialized `World` asset at which the `State` we transition into points
3. Deserialize the new `World`.

We will then have to figure out a new API for events, because if `State`s only load and unload data they can't have a `State::handle_events` method. Also, we will have to figure out how to trigger `State` transitions from `System`s (because `State`s are logic-less) and how to replicate `State::fixed_update` functionality in the new API (if we still want that).

**Comment by Xaeroxe** [comment]
One quick caveat here: audio processing happens roughly 730 times faster than a typical game loop so I do not recommend coupling the two together.  Rodio (our intended audio API) has an audio engine which we could send requests to 60 times a second, but the audio engine itself has to operate independently at a much faster rate.

**Comment by Xaeroxe** [comment]
I'd like to advocate for an exception to the rule that audio output can't happen during the processing phase.  For really simple "play me once in a totally non-spatial way" sounds sending them to a component is unnecessary indirection.  These types of sounds should in my opinion just go straight to the audio engine, as there isn't really any kind of special processing necessary to just start playing them.

**Comment by OvermindDL1** [comment]
I really don't think any direct audio commands should be allowed, rather I'd think they should all go through an intermediary system (that listens to audio 'events' or something) to control the actual back-end, this would allow you do to things like 'delay' sounds based on distance, control how many are allowed to be played at once and at what volumes and so forth.  Nothing really in the engine should talk to the audio backend directly.

**Comment by Xaeroxe** [comment]
Yes that all makes sense, for 3D sounds.  For dead simple 2D sounds such as menu SFX it's just excessive.

**Comment by OvermindDL1** [comment]
> Yes that all makes sense, for 3D sounds. For dead simple 2D sounds such as menu SFX it's just excessive.

Even for dead simple 2d sounds you still want an intermediary layer.  Like take a button click, if it takes 0.3s to play, but the player has a macro program or so and accidentally clicks a thousand times in a second, you want to throttle it.

The audio system I built up was not quite as advanced as the input system and event system and ecs system in it I've described before, but it was very capable.  As I recall you just posted a message to `/dev/sound/<category>` passing in data such as pitch and a few other bits including a set of tags (it was a statically preallocated array that held a max of 10 element) and they applied a few various effects (it also used my fast Atom64 type).  Internally it used those tags to determine whether or not to play in 3d or 2d, if to apply distance calculations (in other words it would not play it instantly but hold on to it for a period of time based on a passed in 'distance/medium' argument), among other effects (some reverb, damping, etc... etc...).

I'd not make it the same way nowadays in terms of capabilities, but the interface was awesome.  Nowadays I'd probably combine it with some auditory reflection libraries for mapping in an environment like that newly released Source library or so.

But even for simple 2d sound, sometimes you still want to apply effects for example (like if you have a game with a 'themed' menu so if you went to a water campaign menu then the menu sounds became distorted like they are under water as one example).

**Comment by Xaeroxe** [comment]
The audio system I'm creating will have effects.  Rodio takes rust iterators as a source, giving you the ability to get in the middle and tweak the samples however you please.  You raise a good point about needing a cap on how many sounds can be played, so sure we'll also delay sending the command until we've processed it in the output phase and determined whether or not we actually want to send it.

**Comment by torkleyy** [comment]
Is there anything written in this issue that we need to implement before closing?

**Comment by Xaeroxe** [comment]
I don't feel the original proposal helps with isolating bugs and it provides what I feel are unnecessary and clumsy restraints to the coding process.

That being said, @OvermindDL1 raised a really good point about limiting the maximum number of sound effects, so I'd like to close this and make a separate issue for that.
